src/modules/git_status.rs

```diff
@@ -219,40 +245,224 @@ fn get_repo_status(
-    let has_untracked = !config.untracked.is_empty();
-    if !has_untracked {
-        args.push("--untracked-files=no");
-    }
-    if config.ignore_submodules {
-        args.push("--ignore-submodules=dirty");
-    } else if !has_untracked {
-        args.push("--ignore-submodules=untracked");
-    }
+        // ... and add flags that omit information the user doesn't want
+        if !has_untracked {
+            args.push("--untracked-files=no");
+        }
+        if config.ignore_submodules {
+            args.push("--ignore-submodules=dirty");
+        } else if !has_untracked {
+            args.push("--ignore-submodules=untracked");
+        }
+
+        let status_output = repo.exec_git(context, &args)?;
+        let statuses = status_output.stdout.lines();
 
-    let status_output = repo.exec_git(context, &args)?;
-    let statuses = status_output.stdout.lines();
+        statuses.for_each(|status| {
+            if status.starts_with("# branch.ab ") {
+                repo_status.set_ahead_behind(status);
+            } else if !status.starts_with('#') {
+                repo_status.add(status);
+            }
+        });
+    } else {
+        let is_interrupted = Arc::new(AtomicBool::new(false));
+        std::thread::Builder::new()
+            .name("starship timer".into())
+            .stack_size(256 * 1024)
+            .spawn({
+                let is_interrupted = is_interrupted.clone();
+                let abort_after =
+                    std::time::Duration::from_millis(context.root_config.command_timeout);
+                move || {
+                    std::thread::sleep(abort_after);
+                    is_interrupted.store(true, std::sync::atomic::Ordering::SeqCst);
+                }
+            })
+            .expect("should be able to spawn timer thread");
+        // We don't show details in submodules.
+        let check_dirty = true;
+        let status = gix_repo
+            .status(gix::features::progress::Discard)
+            .ok()?
+            .index_worktree_submodules(if config.ignore_submodules {
+                Submodule::Given {
+                    ignore: gix::submodule::config::Ignore::Dirty,
+                    check_dirty,
+                }
+            } else if !has_untracked {
+                Submodule::Given {
+                    ignore: gix::submodule::config::Ignore::Untracked,
+                    check_dirty,
+                }
+            } else {
+                Submodule::AsConfigured { check_dirty }
+            })
+            .index_worktree_options_mut(|opts| {
+                // TODO: figure out good defaults for other platforms, maybe make it configurable.
+                //       Git uses everything (if repo-size permits), but that's not the best choice for MacOS.
+                opts.thread_limit = if cfg!(target_os = "macos") {
+                    Some(3)
+                } else {
+                    None
+                };
```

Comment:
I don't have my Windows system at hand right now, but I assume that Windows will also benefit from the lower thread limit. For other operating systems, this should probably be limited to [8 or `STARSHIP_NUM_THREADS`](https://github.com/starship/starship/blob/master/src/main.rs#L284-L290) like the global thread pool to avoid issues on systems with high core counts.